File: python/graph/articulation_points.py
# -*- encoding: utf-8 -*-

# 関節点(Articulation Points)
def get_articulation_points(G, N, start=0):
    v_min = [0]*N; order = [None]*N
    result = []; count = 0
    def dfs(v, prev):
        nonlocal count
        r_min = order[v] = count # 到達時にラベル
        fcnt = 0; p_art = 0
        count += 1
        for w in G[v]:
            if w == prev:
                continue
            if order[w] is None:
                ret = dfs(w, v)
                # 子の頂点が到達できたのが、自身のラベル以上の頂点のみ
                # => 頂点vは関節点
                p_art |= (order[v] <= ret)
                r_min = min(r_min, ret)
                fcnt += 1
            else:
                r_min = min(r_min, order[w])
        p_art |= (r_min == order[v] and len(G[v]) > 1)
        if (prev == -1 and fcnt > 1) or (prev != -1 and p_art):
            # 頂点startの場合は、二箇所以上の子頂点を調べたら自身は関節点
            result.append(v)
        return r_min
    dfs(start, -1)
    return result

File: cpp/graph/articulation_points.cpp
#include<vector>
using namespace std;

// ap[i]: 頂点iが関節点であればtrue

#define N 200005

int n;
vector<int> g[N];
bool ap[N];
int ord[N], low[N], lb = 0;
void dfs(int v, int p) {
  ord[v] = low[v] = lb++;
  int c = 0;
  for(int w : g[v]) {
    if(w == p) continue;
    if(ord[w] == -1) {
      dfs(w, v);
      ++c;
      ap[v] |= (ord[v] <= low[w]);
      low[v] = min(low[v], low[w]);
    } else {
      low[v] = min(low[v], ord[w]);
    }
  }
  if(p == -1) {
    ap[v] = (c > 1);
  }
}

void articulation() {
  for(int i=0; i<n; ++i) {
    ap[i] = false;
  }
  for(int i=0; i<n; ++i) {
    ord[i] = low[i] = -1;
  }
  for(int i=0; i<n; ++i) {
    if(ord[i] != -1) continue;
    dfs(i, -1);
  }
}